Add support for pulling and pushing all images

zubron opened the pull request:
**What this PR does / why we need it**:
This changes adds support for pulling the Sonobuoy worker image, the
systemd-logs image and the Kubernetes conformance image in addition to
the images required for the e2e plugin.

The individual images to pull cannot be configured. Only the default
version for the cluster will be pulled. The images can only be pushed to
one registry which is specified with the new flags `--custom-registry`.

This currently only works for the default built in plugins. A follow up
change will be made to apply the same approach to any image in a plugin
definition.

Signed-off-by: Bridget McErlean <bmcerlean@vmware.com>

**Which issue(s) this PR fixes**
- Fixes #1028 
- Fixes #1059 

**Special notes for your reviewer**:
I've tested the new commands for this manually. I ran the following locally:
```
sonobuoy images pull -p systemd-logs
sonobuoy images push -p systemd-logs --custom-registry zubron
```
And you can see that the images were pushed to my public registry: https://hub.docker.com/u/zubron

**Release note**:
```
Sonobuoy can now pull all images for Sonobuoy and the built-in plugins and push them to a custom registry. The E2E test images are still configured as before, however all other images will be pushed to a custom registry specified using the flag `--custom-registry`. The image operations can also be previewed by using the `--dry-run` flag.
```

zubron commented on cmd/sonobuoy/app/images.go:
```diff
@@ -165,124 +176,180 @@ func getClusterVersion(kubeconfig Kubeconfig) (string, error) {
+func downloadImages(plugins []string, kubeconfig Kubeconfig, e2eRegistryConfig string) error {
+	for _, plugin := range plugins {
+		switch plugin {
+		case e2ePlugin:
```
I haven't updated this function as I don't know if we should maintain it: #1058

zubron commented on cmd/sonobuoy/app/images.go:
```diff
@@ -165,124 +176,180 @@ func getClusterVersion(kubeconfig Kubeconfig) (string, error) {
-		return imageClient.DeleteImages(images, numDockerRetries)
+func substituteRegistry(image string, customRegistry string) string {
+	trimmedRegistry := strings.TrimRight(customRegistry, "/")
+	components := strings.SplitAfter(image, "/")
```
I thought I left a comment on this but it's gotten lost :confused: 

This is a best attempt at replacing the registry in a full image name. It's possible for the "name" component of an image name to also contain a `/` and we see that in [some of the E2E test images](https://github.com/vmware-tanzu/sonobuoy/blob/master/pkg/image/v1.17.go#L71-L74). In cases like these, this function would substitute everything up to the last `/` which could include the registry *and* part of the image name. I don't think there's a way to identify what is the registry URL component of a FQIN. The images will still tag and push correctly, but it might be confusing for users when replacing images in their plugin definitions.

johnSchnake replied:
Its a short function but I think this is a good function to have in a function so that we can visually ensure it handles certain cases.

zubron commented on cmd/sonobuoy/app/args.go:
```diff
@@ -348,6 +351,11 @@ func AddPluginEnvFlag(p *PluginEnvVars, flags *pflag.FlagSet) {
 
+// AddPluginListFlag adds the flag to keep track of which built-in plugins to use.
+func AddPluginListFlag(p *[]string, flags *pflag.FlagSet) {
+	flags.StringSliceVarP(p, "plugin", "p", []string{"e2e"}, "Describe which plugin's images to interact (Valid plugins are 'e2e', 'systemd-logs').")
```
Should this default to both plugins now? Seems like that would be simpler UX here.

johnSchnake replied:
Yes it looks like this is just used for the images command. If we are going to, by default, try to run both of those plugins and the `images` commands can support both of them, I agree it should.

johnSchnake commented on cmd/sonobuoy/app/images.go:
```diff
@@ -165,124 +176,180 @@ func getClusterVersion(kubeconfig Kubeconfig) (string, error) {
-		if err != nil {
-			return errors.Wrap(err, "failed to get cluster version")
-		}
+func listImages(plugins []string, kubeconfig Kubeconfig) error {
```
This is another good candidate for a test function; we can add it when we expand to other plugins though. I know the systemd-logs test wouldnt be very significant and the e2e one is a bit awkward to test.

johnSchnake commented:
My only other idea for this is if it would be useful to have a `dry-run` option when pushing so that you can confirm the repos/tags are as expected.

When trying to manually confirm this all works, I can't actually check much manually without _actually_ pushing some images.

Otherwise, my only comments were (already commented but adding for clarity):
 - ideally add another test to ensure we are splitting registries/tags appropriately
 - default to doing both systemd-logs and e2e

zubron commented:
Thanks for the feedback! I've addressed your two comments. I also added a `dry-run` option which just performs the logging operations that the existing docker image client does. I've left it as a separate commit for now. Let me know what you think!
